### PR TITLE
Add tapioca compiler for Homebrew::CLI::Args

### PR DESCRIPTION
**Pull request by dduugg**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----
See context at https://github.com/Homebrew/brew/pull/16815#issuecomment-1990530583

I wanted to put this off a bit, since currently all command args live in the same namespace. That means that any arg accepted by any command will typecheck when used in any other command. Thus, this PR adds an RBI that allows 297 methods on `Homebrew::CLI::Args`. 

However, this should allow us to roll out the `AbstractCommand` interface without resorting to hash accessors or weakening the contract that the interface provides.

I'm opening this up as a draft for initial feedback. I will see how it integrates with the PR linked above as time allows in the coming days.

### CODE REVIEW

**Comment by issyl0** [comment]
This gives me inspiration and avenues to debug my current non-working branch for a Tapioca compiler for Tty, to replace the generator script, so thank you!

**Comment by dduugg** [comment]
Ok, I've updated https://github.com/Homebrew/brew/pull/16815 to validate that this approach removes the need to make changes to either Library/Homebrew/cli/args.rb or how `args` is accessed within commands in that PR. Marking this PR as r4r. 🎉

File: Library/Homebrew/sorbet/tapioca/compilers/args.rb [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+module Tapioca
+  module Compilers
+    class Args < Tapioca::Dsl::Compiler
+      # This is ugly, but we're moving to a new interface that will use a consistent DSL
+      # These are cmd/dev-cmd methods that end in `_args` but are not parsers
+      NON_PARSER_ARGS_METHODS = T.let([
+        :formulae_all_installs_from_args,
+        :reproducible_gnutar_args,
+        :tar_args,
```
[the remark this answers is not in the document]
Comment (dduugg) [reply Library/Homebrew/sorbet/tapioca/compilers/args.rb:19]: It should be nearly as fast to just port the impacted methods to use `AbstractCommand` (🤞)

File: Library/Homebrew/sorbet/tapioca/compilers/args.rb [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+      def decorate
+        root.create_path(Homebrew::CLI::Args) do |klass|
+          Homebrew.methods(false).select { _1.end_with?("_args") }.each do |args_method_name|
+            next if NON_PARSER_ARGS_METHODS.include?(args_method_name)
+
+            parser = Homebrew.method(args_method_name).call
+            comma_array_methods = comma_arrays(parser)
+            args = parser.instance_variable_get(:@args)
+            args.instance_variable_get(:@table).each do |method_name, value|
+              # some args are used in multiple commands (this is ok as long as they have the same type)
```
Comment (dduugg) [inline Library/Homebrew/sorbet/tapioca/compilers/args.rb:40]: We could add some validation that method signatures are unchanged, but this one be a problem once we have namespaced args

File: Library/Homebrew/sorbet/tapioca/compilers/args.rb [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+      sig { override.void }
+      def decorate
+        root.create_path(Homebrew::CLI::Args) do |klass|
+          Homebrew.methods(false).select { _1.end_with?("_args") }.each do |args_method_name|
+            next if NON_PARSER_ARGS_METHODS.include?(args_method_name)
+
+            parser = Homebrew.method(args_method_name).call
+            comma_array_methods = comma_arrays(parser)
+            args = parser.instance_variable_get(:@args)
+            args.instance_variable_get(:@table).each do |method_name, value|
```
Comment (dduugg) [inline Library/Homebrew/sorbet/tapioca/compilers/args.rb:39]: We're pretty deep in the internals of the parser at this point (accessing the ivar of an ivar), it might be a good idea to write tests and/or update CLI::Parser to more directly expose its method table.

File: Library/Homebrew/sorbet/rbi/upstream.rbi [hunk 4]
```diff
@@ -2,21 +2,3 @@
 
 # This file contains temporary definitions for fixes that have
 # been submitted upstream to https://github.com/sorbet/sorbet.
-
-# https://github.com/sorbet/sorbet/pull/7650
```
Comment (dduugg) [inline Library/Homebrew/sorbet/rbi/upstream.rbi:6]: unrelated cleanup, while I'm in the `sorbet` dir (the linked upstream PRs have merged)

File: Library/Homebrew/sorbet/tapioca/compilers/args.rb (hunk 3 above)
Comment (apainintheneck) [reply Library/Homebrew/sorbet/tapioca/compilers/args.rb:39]: Those sound like good ideas. If we expose `:@table`, we should make sure that it's read only.

**Review by apainintheneck — APPROVED** [review]

File: Library/Homebrew/sorbet/tapioca/compilers/args.rb (hunk 3 above)
Comment (dduugg) [reply Library/Homebrew/sorbet/tapioca/compilers/args.rb]: Ok, I broke up the logic in order to test various aspects of this compiler. (I decided not make any changes to CLI::Parser). PTAL.

File: Library/Homebrew/sorbet/tapioca/compilers/args.rb [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+      NON_PARSER_ARGS_METHODS = T.let([
+        :formulae_all_installs_from_args,
+        :reproducible_gnutar_args,
+        :tar_args,
+      ].freeze, T::Array[Symbol])
+
+      # FIXME: Enable cop again when https://github.com/sorbet/sorbet/issues/3532 is fixed.
+      # rubocop:disable Style/MutableConstant
+      ConstantType = type_member { { fixed: T.class_of(Homebrew::CLI::Args) } }
+      # rubocop:enable Style/MutableConstant
```
Comment (carlocab) [inline Library/Homebrew/sorbet/tapioca/compilers/args.rb:20]: I don't follow where you use this -- care to elaborate?

Comment (dduugg) [reply Library/Homebrew/sorbet/tapioca/compilers/args.rb:20]: Per Tapioca [docs](https://github.com/Shopify/tapioca?tab=readme-ov-file#writing-custom-dsl-compilers):

> Every compiler must declare the type member ConstantType in order for Sorbet to understand what the return type of the constant attribute reader is. It needs to be assigned the correct type variable matching the type of constants that gather_constants returns. This generic variable allows Sorbet to type-check method calls on the constant reader in your decorate method. See the Sorbet documentation on [generics](https://sorbet.org/docs/generics) for more information.

**Comment by dduugg** [comment]
I believe feedback has been addressed, I'm going to smash that green button and resume work on the related [PR](https://github.com/Homebrew/brew/pull/16815). Feel free to comment on that one if I've missed anything here, and I'll address over there. 🙇‍♂️

**Review by MikeMcQuaid — COMMENTED** [review]
Looks good, thanks @dduugg!
